File: samples/jpeg_encode/jpeg_encode_prof.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <sys/stat.h>

#include <fstream>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/jencoder.hpp"
#include "common/media_profiler.hpp"
#include "common/utils.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("device_ids", 'd', "device id to run", false, "[0]");
  args.add<std::string>("input_file", '\0', "input file", false,
                        "../data/images/plate_1920_1080.yuv");

  args.add<uint32_t>("width", '\0', "width", false, 1920);
  args.add<uint32_t>("height", '\0', "height", false, 1080);
  args.add<uint32_t>("instance", 'i', "instance number for each device", false,
                     1);
  args.add<int>("iterations", '\0', "iterations count for one profiling", false,
                10240);
  args.add<std::string>("percentiles", '\0', "percentiles of latency", false,
                        "[50, 90, 95, 99]");

  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);

  std::string input_file = args.get<std::string>("input_file");
  uint32_t width = args.get<uint32_t>("width");
  uint32_t height = args.get<uint32_t>("height");
  auto device_ids = vsx::ParseVecUint(args.get<std::string>("device_ids"));
  auto instance = args.get<uint32_t>("instance");
  auto iterations = args.get<int>("iterations");
  auto percentiles = vsx::ParseVecUint(args.get<std::string>("percentiles"));

  std::vector<std::shared_ptr<vsx::Jencoder>> jencoders;
  jencoders.reserve(instance);
  for (uint32_t i = 0; i < instance; i++) {
    uint32_t device_id = device_ids[i % (device_ids.size())];
    vsx::SetDevice(device_id);
    jencoders.push_back(std::make_shared<vsx::Jencoder>(
        device_id, input_file, width, height, vsx::ImageFormat::YUV_NV12));
  }

  vsx::MediaProfilerConfig config = {instance, iterations, device_ids,
                                     percentiles};
  vsx::MediaProfiler<vsx::Jencoder, std::shared_ptr<vsx::DataManager>> profiler(
      config, jencoders);
  std::cout << profiler.Profiling() << std::endl;
  return 0;
}

File: samples/jpeg_encode/jpeg_encode_prof.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys
import vaststreamx as vsx

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from pathlib import Path

from common.jencoder import Jencoder
from common.media_profiler import MediaProfiler
from easydict import EasyDict as edict
import argparse
import ast


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_ids",
        default="[0]",
        help="device ids to run",
    )
    parser.add_argument(
        "--width",
        default=0,
        type=int,
        help="frame width",
    )
    parser.add_argument(
        "--height",
        default=0,
        type=int,
        help="frame height",
    )
    parser.add_argument(
        "--input_file",
        default="../data/images/plate_1920_1080.yuv",
        help="input file path",
    )
    parser.add_argument(
        "-i",
        "--instance",
        default=1,
        type=int,
        help="instance number for each device",
    )
    parser.add_argument(
        "--iterations",
        default=10240,
        type=int,
        help="iterations count for one profiling",
    )
    parser.add_argument(
        "--percentiles",
        default="[50, 90, 95, 99]",
        help="percentiles of latency",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    device_ids = ast.literal_eval(args.device_ids)
    instance = args.instance
    iterations = args.iterations
    percentiles = ast.literal_eval(args.percentiles)

    input_file = args.input_file

    encoders = []
    for i in range(instance):
        device_id = device_ids[i % len(device_ids)]
        encoder = Jencoder(
            device_id,
            input_file,
            args.width,
            args.height,
            vsx.ImageFormat.YUV_NV12,
        )
        encoders.append(encoder)

    # print(f"len encoder: {len(encoders)}")
    config = edict(
        {
            "instance": instance,
            "iterations": iterations,
            "device_ids": device_ids,
            "percentiles": percentiles,
        }
    )
    profiler = MediaProfiler(config, encoders)
    print(profiler.profiling())
